code for Fifo_module

Siddharth504 opened the pull request:
Link for Fifo_module Block Diagram
https://www.lucidchart.com/publicSegments/view/30b1ea21-d68a-408c-835b-f093e6327303

Siddharth504 committed:
code for Fifo_module

VlsiCommonCode reviewed (commented):
push and pop has to be part of sensitivity list at line 23
why data out is blocking? 54:57 looks wrong to me, data has to traverse from the bottom up right? 
Can you explain me the case at line 55 statement what you are trying to achieve with this? 
fifo_empty and fifo_full have to be output instead of reg?

Siddharth504 commented:
> push and pop has to be part of sensitivity list at line 23
> why data out is blocking? 54:57 looks wrong to me, data has to traverse from the bottom up right?
> Can you explain me the case at line 55 statement what you are trying to achieve with this?
> fifo_empty and fifo_full have to be output instead of reg?
push and pop has to be the part of sensitivity list : 
No, I doesn't think it should be in sensitivity list. At line 23 always block is used for sequential circuit consisting of registers which is mainly collection of flip flop.So in always block there should be presence of clk in sensitivity list  not push and pop.
We can use push and pop in sensitivity list when always block is used for combination logic like at line 49 and 67.

Siddharth504 commented:
> push and pop has to be part of sensitivity list at line 23
> why data out is blocking? 54:57 looks wrong to me, data has to traverse from the bottom up right?
> Can you explain me the case at line 55 statement what you are trying to achieve with this?
> fifo_empty and fifo_full have to be output instead of reg
Explanation: why data out is blocking? 54:57 looks wrong to me, data has to traverse from the bottom up right?
we use non- blocking assignment so that all the statement in a particular block runs together at a time without blocking  another statement from being executed while in blocking assignment second statement is not executed unless first statement is executed. 
I had used data_out in the case statement block with blocking assignment. But we know  in the case statement only one statement is selected at a time depending on the condition inside   "case(condition)".
so here there is no any situation arises that two statement should run together or one after another because at a time only one particular statement is choosen  and executed. So according to me selecting data_out as a blocking or non-blocking assignment doesn't matter much.
Here as only  one  particular statement depending on the condition so I think there is not any question for traversing data from bottom up .

Siddharth504 commented:
> push and pop has to be part of sensitivity list at line 23
> why data out is blocking? 54:57 looks wrong to me, data has to traverse from the bottom up right?
> Can you explain me the case at line 55 statement what you are trying to achieve with this?
> fifo_empty and fifo_full have to be output instead of reg?
Can you explain me the case at line 55 statement what you are trying to achieve with this?
Explanation: count is the select line for the mux .
Depending upon the value of count ,data_out gets the data stored in the register. At line 55 when the count value is 2 then the statement at the line 55 is executed and data _out get the value stored in the register that is data_mux2 .  data_mux2 is the value stored in the second register(from the block diagram its the second register from the top).later  its also shown in diagram that 4 values stored in 4 different register acts as 4 input to mux and depending on select line(count) data_out gets a particular value.

Siddharth504 commented:
> push and pop has to be part of sensitivity list at line 23
> why data out is blocking? 54:57 looks wrong to me, data has to traverse from the bottom up right?
> Can you explain me the case at line 55 statement what you are trying to achieve with this?
> fifo_empty and fifo_full have to be output instead of reg?
Explanation:  fifo_empty and fifo_full have to be output instead of reg?
yes fifo_empty and fifo_full is the output. That is declared in the code.But I have also used fifo_empty and fifo_full in the always block.I have studied that output if used inside always block  then its needs to be declare reg.
We also use reg so that variable hold its value until another value is placed on them.In the code we had used fifo_full and fifo_empty to store two different values at two different time and reflect it as a output, so its possible only if we use reg because "wire" declaration doesn't store value .Wire only represents physical connection between components.
Also when we remove reg declaration from fifo_full and fifo_empty it shows compile  error in the code.

Siddharth504 committed:
updated fifo code 

Used  push or pop in sensitivity list at line 49 and count in sensitivity list at line 67

VlsiCommonCode commented:
Overall code looks good to me and I am convince with your feedback apart from data traversing I still believe data is wrong :)
Couple of thing further which you can do to complete design verification  flow

1. Think of verification scenario (test cases) to verify your design 
2. Please attach snapshot of waveform as a reference of functionality
